x/programs/rust/README.md

```diff
@@ -25,6 +25,100 @@ Just a couple things to note. Serialization is minimal, yet there are certain as
 
 On the Go side, we unmarshal in the same order.
 
+## Storage
+
+A Program has 2 storage types.
+
+- `Memory` in WebAssembly is a linear buffer of unsigned bytes that can read and
+written to by both the guest or host. Because of the very minimal types
+supported by WebAssembly Memory is critical for communication between `Program`
+and host.
+
+- `State` is a host module which is imported into the `Program` instance, which
+exposes access to a database store. Unlike `Memory` state can be durable.
+
+### Memory
+
+In this example we share a byte slice between our host and a `Program`. To
+achieve this we can write our byte slice to the shared memory from he host using
+an exported functions called `alloc` and `memory` To do this we would first
+allocate  a chunk of continuous memory of the length desired.  `Alloc` will
+return the `offset` of that slice.
+
+
+```go
+runtime := // create new runtime and initialize
+memory := runtime.Memory()
+
+bytes := []byte("hello world")
+bytesLen := uint64(len(bytes))
+
+// create a memory allocation 
+offset, _ := memory.Alloc(bytesLen)
+
+// copy bytes into memory allocation
+offset, _ = memory.Write(offset, bytes)
+
+// pass the offset and lenth as params to a function
+// exported by the Program
+resp, _ := runtime.Call("some_function", offset, bytesLen )
+```
+
+`hello world` converted to a byte slice will represent our Program's memory for
+the example
+```sh
+[offset=6              ][length=3   ]
+[104 101 108 108 111 32 119 111 114 108 100]
+[119 111 114] // wor
+```
+The below shows an exported `Program` function `some_function` which accepts a
+message of arbitrary length. Unlike fixed size length objects such as `Address`
+the length can be variable. So the `length` of the memory along with the
+`offset` are both required.
+
+```rust
+#[public]
+pub fn some_function(program: Program, message: i64, message_length: i64) -> i64 {
+  // read message bytes from memory
+  let msg := program.memory().read(message, message_length)?;
+  // msg = [119 111 114]
+}
+```
+          
+Each  Memory is tightly bound to the lifecycle of the module. For this reason
+`Memory` is non durable and can not be relied upon. 
+
+### State
+
+- `Storage` is the definition of the schema used to persist Program data.
+
+```rust
+#[reps(u8)]
+#[storage]
```

Comment:
🤔, the `storage` attribute should automatically add a `repr` attribute. I'm guessing that `repr(Rust)` would be invalid here, so the `Storage(u8)` could validate its input and pass it to `repr`

Reply:
makes sense

Reply:
@hexfusion, FYI. I stumbled upon this if you're looking for inspiration
https://docs.rs/enumflags2_derive/latest/src/enumflags2_derive/lib.rs.html#56-79

They force you to add `repr` separately, but I think the validation would work the same way

Reply:
Check out the state macro I added it explicitly. lmk what you think

https://github.com/ava-labs/hypersdk/pull/499/files#diff-3ca848a15e3cf3a85a35434debabf380e80ce7a22ef64446e861cb2b93c9d1c1R96